x/crosschain/keeper/msg_server_vote_outbound.go

```diff
@@ -136,18 +138,104 @@ func (k msgServer) VoteOutbound(
+func (k Keeper) RefundUnusedGasFee(ctx sdk.Context, cctx *types.CrossChainTx) {
+	outboundParams := cctx.GetCurrentOutboundParam()
+	// We skip funding the gas stability pool if the userGasFeePaid is nil or zero.
+	// This is a legacy outbound where the user fee was not recorded as part of the cctx struct. This is only to handle cctxs which might be in pending outbound state when the upgrade happens and get finalized after the upgrade
+	if outboundParams.UserGasFeePaid.IsNil() {
+		err := k.FundGasStabilityPoolFromRemainingFees(
+			ctx,
+			*outboundParams,
+			outboundParams.ReceiverChainId,
+		)
+		if err != nil {
+			ctx.Logger().Error("Failed to fund gas stability pool (Legacy) with remaining fees",
+				"voteOutboundID", voteOutboundID,
+				"cctxIndex", cctx.Index,
+				"error", err,
+			)
+		}
+		return
+	}
+
+	if err := k.refundUnusedGas(ctx, cctx); err != nil {
+		ctx.Logger().Error("Failed to fund gas stability pool with remaining fees",
+			"voteOutboundID", voteOutboundID,
+			"cctxIndex", cctx.Index,
+			"error", err,
+		)
+	}
+}
+
+// RefundUnusedGas uses the remaining fees of an outbound tx to fund the gas stability pool and refund the user if possible
+func (k Keeper) refundUnusedGas(
+	ctx sdk.Context,
+	cctx *types.CrossChainTx,
+) error {
+	outboundParams := cctx.GetCurrentOutboundParam()
+	outboundTxFeePaid := math.NewUint(outboundParams.GasUsed).
+		Mul(math.NewUintFromBigInt(outboundParams.EffectiveGasPrice.BigInt()))
+	userGasFeePaid := outboundParams.UserGasFeePaid
+
+	// Handle cases for not funding the stability pool or refunding the user
+	// 1. If outboundTxFeePaid is nil or zero, we do not fund the stability pool or refund the user.This is for Non EVM chains which do not populate outboundTxFeePaid
+	// 2. If outboundTxFeePaid is greater than or equal to userGasFeePaid, we do not fund the stability pool or refund the user. Since the outbound tx used all the gas paid by the user.The additional gas used is covered by the stability pool.
+	// https://github.com/zeta-chain/node/issues/4219
+	// Enable for non EVM chains once zeta-client supports populating gas used and effective gas price for non EVM chains.
+	if outboundTxFeePaid.IsNil() || outboundTxFeePaid.IsZero() || outboundTxFeePaid.GTE(userGasFeePaid) {
+		return nil
 	}
+
+	// We use a maximum for 95 percent of the remaining fees to fund the stability pool and refund the user.
+	usableRemainingFees := PercentOf(
+		outboundParams.UserGasFeePaid.Sub(outboundTxFeePaid),
+		types.UsableRemainingFeesPercentage,
+	)
+	if !usableRemainingFees.GT(math.ZeroUint()) {
+		return nil
+	}
+
+	// Send all tokens to stability pool by default
+	stabilityPoolPercentage := uint64(100)
+	refundToUser := false
+
+	// Refund to the sender irrespective of whether its EOA or contract address if it's a withdrawal originating from zEVM.
+	if chains.IsZetaChain(cctx.InboundParams.SenderChainId, k.GetAuthorityKeeper().GetAdditionalChainList(ctx)) &&
```

Comment:
We should double check, I think we have a utility function with the cctx object to check it's send from ZetaChain? Instead of doing this check manually